ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+  options)
+
+
+## Defining a workflow
+
+Similar to activities, workflows are also represented as classes.
+
+`sum_workflow.rb`
+
+```ruby
+require 'sum_activity'
+
+# Similar to activities the class name will be used as a workflow name unless overridden
+class SumWorkflow < Temporal::Workflow
+  # Optionally override the name
+  workflow_name 'sum-workflow'
+
+  def execute(a, b)
+    # - Positional arguments passed to activity (inline with other SDKs)
+    # - Keyword arguments are reserved for options
+    # - This call is blocking
+    workflow.execute_activity(SumActivity, a, b, start_to_close_timeout: 10)
+
+    # - Starting an activity by its name rather than a class
+    # - The result of this activity will be used as a workflow result (Ruby standard)
+    workflow.execute_activity('sum-activity', a, b, start_to_close_timeout: 10)
+  end
+end
+```
+
+#### Alternative approaches
+
+- For the same exact reasons it is not feasible to use non-namespaced methods for workflow
+
+- Similar to activities it should be possible to define multiple workflows per class, however this
+  will not be allowed to avoid a perception of a shared state between workflows and allow for
+  class-level "decorators"
+
+
+## Start a workflow
+
+```ruby
+require 'temporal/client'
+require 'sum_workflow'
+
+# Connections are eager and a lazy option will be introduced later
+connection = Temporal::Connection.connect('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+# workflow is referenced by the class
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+handle.result # => 3
+```
+
+#### Alternative approaches
+
+- Starting a foreign workflow from a string name
+
+```ruby
+client.start_workflow('sum-workflow', 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+- Starting a foreign workflow with the help of a stub
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  # Optionally override the name of the workflow
+  workflow_name 'my.workflow'
+end
+
+client.start_workflow(MyWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+
+## Start a worker
+
+```ruby
+require 'temporal/worker'
+require 'sum_workflow'
+require 'some_other_workflow'
+require 'sum_activity'
+require 'some_other_activity'
+
+connection = Temporal::Connection.connect('localhost:7233')
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [SumWorkflow, SomeOtherWorkflow],
+  activities: [SumActivity, SomeOtherActivity]
+)
+
+worker.run
+```
+
+#### Alternative approach
+
+To support passing additional context to the activity classes a worker can be initialised with an
+instance of the activity class (only available for activities).
+
+```ruby
+redis = Redis.new('localhost:6379')
+
+Temporal::Worker.new(
+  ...,
+  activities: [CalculatorActivities.new(redis)]
+)
+```
+
+*NOTE: This however will come with a heavy warning — the activity must be strictly thread-safe as
```

Comment:
This warning should probably exist for _all_ activities, not just ones they instantiate. Because we should be instantiating the class-activities at registration time too. This is so the user can use (thread-safe) fields of the class across activities if they so choose.

Basically, the lifetime of all activity classes should be at least the worker level, whether we instantiate them or they do. It'd be confusing if class state was per-activity for one way of registering and per-worker for another.

Reply:
Wondering we should require the user to instantiate their own classes, it eliminates questions of "when / where is this instantiated" and makes it so there's a single way to register activities.

Reply:
If we expect the Worker to instantiate client provided activity classes at the registration time then there is no need for accepting classes and we can just accept instances. My biggest worry is that we're exposing every single Ruby SDK user to the thread-safety issues that they might not be very familiar with (Ruby is known for not being 100% thread safe, which is the main reason for the Global Interpreter Lock).

By accepting a class rather than an instance I was hoping that we can create a new Activity instance only on a dedicated thread when processing an activity task therefore making it much easier and safer for the majority. If we don't care about this then I'm happy with only accepting instances (I don't like the discrepancy with workflows, but I can probably live with it)

Reply:
I think accepting classes makes plenty of sense and reads better, with the option of accepting instances too for advanced use.